Adding a service (#184)
madewithkoji/docs: src/asciidoc/docs/develop/addService.adoc

There are several values your app will need to get from Koji.

* The URL for the frontend.
* The URL for the backend.
* The project ID

Koji stores these values in internal variables. To make them accessible to your app, you need to map them to environment variables, and then add them to the configuration.

. Open the frontend/.env file, update the contents as follows, and then save the file.
+
[source,text]
----
REACT_APP_FRONTEND_URL=$KOJI_SERVICE_URL_frontend
REACT_APP_BACKEND_URL=$KOJI_SERVICE_URL_backend
REACT_APP_PROJECT_ID=$KOJI_PROJECT_ID
----
. Open the frontend/src/index.js file, update the config call as follows, and then save the file.
+
[source,javascript]
----
....
Koji.config(kojiConfig, {
  projectId: process.env.REACT_APP_PROJECT_ID,
  services: {
    backend: process.env.REACT_APP_BACKEND_URL,
    frontend: process.env.REACT_APP_FRONTEND_URL
  },
});
....
----

== Letting Koji know about the backend service

You have now added a backend project, but the Koji editor and app builder still don't know it exists.
You must update the `koji.json` file to let Koji know about the new service.
`koji.json` is located in the root directory.

> . Open the `koji.json` file, and replace the existing content with the following code.

Reviewer: Should we focus these instructions on the develop and deploy keys? I believe that’s what a developer needs to update for this tutorial. If so, believe you can use the same includes that I used for those sections in the Project structure and configuration page. 

Something like, … 

```suggestion
. Open the `koji.json` file, and add the deployment and development instructions for the backend service to the `deploy` and `develop` keys.
+
[source,JSON]
----
include::{includespath}/koji.json[tags=develop;deploy,indent=0]
----
--
```

+
[source,json]
----
{
  "deploy": {
    "frontend": {
      "output": "frontend/build",
      "type": "static",
      "commands": [
        "cd frontend",
        "npm install",
        "export NODE_ENV=production && npm run build"
      ]
    },
    "backend": {
      "output": "backend",
      "type": "dynamic",
      "commands": [
        "cd backend",
        "npm install",
        "export NODE_ENV=production && npm run compile"
      ]
    }
  },
  "develop": {
    "frontend": {
      "path": "frontend",
      "port": 8080,
      "startCommand": "npm start"
    },
    "backend": {
      "path": "backend",
      "port": 3333,
      "startCommand": "npm run start-dev"
    }
  },
  "entitlements": {
    "InstantRemixing": true,
    "InstantRemixingNativeNavigation": true,

>     "FeedEvents": true

Reviewer: Remove `FeedEvents`. This entitlement was deprecated.

Author: This is part of the scaffold. It would have to be removed from there to remove it from this app, unless you wanted to direct the reader to remove it. Otherwise, no reference to it will appear in the doc, since I'm using the include file instead of listing the entire contents of koji.json.

Author: This is part of the scaffold. It would have to be removed from there to remove it from this app, unless you wanted to direct the reader to remove it. Otherwise, no reference to it will appear in the doc, since I'm using the include file instead of listing the entire contents of koji.json.

Reviewer: No, we shouldn't need to remove it from the scaffold. It shouldn't break anything (it just won't do anything). We just don't want it in the docs. It sounds like this won't be an issue since you made the previous change. Thanks!

  },
  "remixData": {
    "backgroundColor": "#282c34"
  },
  "@@initialTransform": {
    "remixData": {}
  }
}
----
+
This file lets the Koji editor know which services to run, so you must add entries for the `backend` service.
. On the left side of the editor, go to *Advanced > Remote environment* and click *Force restart project*.
+
If everything went well, you should be able to reconnect to the editor and see that a `backend` terminal has been added at the bottom.
+
If you see `[koji] backend started`, your backend service is now running in your local editor environment.

== Testing the backend service

To test that everything is working correctly, you can modify your frontend code to get a value from the backend service and display it automatically.

> . Go to the `frontend` folder, press CTL-C to stop the frontend server, and install the `styled-components` package, which the new code will use.

Reviewer: We recently enabled some Asciidoc macros for keyboard instructions, buttons, and menus. As you update pages, can you start applying them to anything you see? Here’s the docs for the macros: https://docs.asciidoctor.org/asciidoc/latest/macros/keyboard-macro/, https://docs.asciidoctor.org/asciidoc/latest/macros/ui-macros/ Let me know if you have questions about how they work.

Also, I would make Install … a separate step.

```suggestion
. In the `frontend` terminal, press kbd:[CTRL+C] to stop the frontend server.
. Install the `styled-components` package, which the new code will use.
```

+
[source,bash]
----
npm install styled-components
----
. Restart the frontend server.
+
[source,bash]
----

> npm run start-dev

Reviewer: Isn’t this the command to start the backend server? If we have them use the frontend terminal to install the styled-components package, I believe we can change this line to `npm start`.

. Open the live preview and refresh it.
+
If everything worked, you should see the first text state for 5 seconds, and then it should be replaced with the `Hello, world!` text from the server.

> . Publish your template and confirm that it's working as expected.

Reviewer: ```suggestion
. Publish your app and confirm that it's working as expected.
```
